### PR TITLE
DSE restructuring and improvements

### PR DESCRIPTION
**Pull request by FabioLuporini**
You'll see lots of edits in this PR, but no worries: it's mostly just moving stuff around and dropping useless functionalities.

Now the DSE has its own subdirectory ``dse``, and its components are split over multiple files.

The Bhaskara polynomials approximation is now applicable by the DSE; as such, it's been removed from TTI.

A few more things:
- Now we always print the compilation time
- The ``factorized`` attribute has finally been dropped 
- Added basic infrastructure for creating ``temporaries graphs``, which will be used in subsequent versions of the DSE

All new functionalities are properly tested

I'm happy to rebase if #161 lands first

### CODE REVIEW

File: devito/dse/inspection.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        return expr.func(*expr.atoms())
+    elif isinstance(expr, Indexed):
+        return expr.func(*expr.args)
+    elif expr.is_Symbol:
+        return expr.func(expr.name)
+    elif expr in [S.Zero, S.One, S.NegativeOne, S.Half]:
+        return expr.func()
+    elif expr.is_Atom:
+        return expr.func(*expr.atoms())
+    if expr.is_Add:
```
Comment (mlange05) [inline devito/dse/inspection.py]: `if` -> `elif`?

File: devito/dse/symbolics.py [hunk 2]
```diff
@@ -125,18 +73,21 @@ class Rewriter(object):
...
-        if mode in ['basic', 'advanced']:
-            processed = self._cse()
+        if mode.intersection({'basic', 'advanced'}):
+            processed = self._cse(processed)
 
-        if mode in ['advanced']:
+        if mode.intersection({'factorize', 'advanced'}):
             processed = self._factorize(processed)
 
+        if mode.intersection({'approx-trigonometry', 'advanced'}):
```
Comment (mlange05) [inline devito/dse/symbolics.py]: Should the new modes be explained in the `rewrite()` docstring?

File: tests/test_rewriter.py [hunk 3]
```diff
@@ -76,17 +80,46 @@ def tti_operator(dse=False):
     return handle
 
 
-def test_tti_rewrite_basic():
-    output1 = tti_operator(dse=None).apply()
-    output2 = tti_operator(dse='basic').apply()
+@pytest.fixture(scope="session")
+def tti_nodse():
+    # FIXME: note that np.copy is necessary because of the broken caching system
```
Comment (mlange05) [inline tests/test_rewriter.py]: Is this still true?

File: devito/dse/inspection.py (hunk 1 above)
Comment (FabioLuporini) [inline devito/dse/inspection.py]: good spot

File: devito/dse/symbolics.py (hunk 2 above)
Comment (FabioLuporini) [inline devito/dse/symbolics.py]: Right, forgot about that

File: tests/test_rewriter.py (hunk 3 above)
Comment (FabioLuporini) [inline tests/test_rewriter.py]: Maybe not, I will try !

**Comment by FabioLuporini** [comment]
Will rebase and address your comments soon

**Comment by mloubout** [comment]
Please make sure that any `tti` example/test uses `space_order=4,8,10..`

**Comment by FabioLuporini** [comment]
Rebased.

If Travis is happy, we can merge this.

File: devito/dse/symbolics.py (hunk 2 above)
Comment (FabioLuporini) [inline devito/dse/symbolics.py]: Done

File: tests/test_rewriter.py (hunk 3 above)
Comment (FabioLuporini) [inline tests/test_rewriter.py]: Forgot to try this. I can handle this in the next PR though (as I might have to change that bit of code anyway).

**Pull request merged**